File: 백준/Gold/17298. 오큰수/오큰수.py
import sys
input = sys.stdin.readline

n = int(input().rstrip())
A = list(map(int, input().rstrip().split()))
stack = []
nge = [-1 for _ in range(n)]
for i in range(n-1, -1, -1):
    # stack이 비어있다면 -1 & push
    if not stack:
        stack.append(A[i])
        continue

    while stack:
        # stack의 top이 현재 숫자보다 작거나 같으면 pop
        # pop한 후 stack이 비게 되면 push하고 다음으로
        if stack[-1] <= A[i]:
            stack.pop()
            if not stack:
                stack.append(A[i])
                break
        # stack의 top이 현재 숫자보다 크다면 nge에 해당 숫자 기입 및 push, break
        else:
            nge[i] = stack[-1]
            stack.append(A[i])
            break

print(' '.join(map(str, nge)))

File: 백준/Gold/17298. 오큰수/오큰수.java
import java.util.*;
import java.io.*;

public class Main {
    public static void main(String[] args) throws IOException {
        new Main().solution();
    }

    public void solution() throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        int n = Integer.parseInt(br.readLine());
        int[] numbers = Arrays.stream(br.readLine().split(" "))
                .mapToInt(Integer::parseInt).toArray();
        int[] answers = new int[n];
        Stack<Integer> stack = new Stack<>();

        for (int i = n-1; i >= 0; i --) {
            int currentElem = numbers[i];
            // 비어있지 않고, 현재 원소가 top보다 크면 pop
            while (!stack.isEmpty() && currentElem >= stack.peek()) {
                stack.pop();
            }
            if (stack.isEmpty()) {
                answers[i] = -1;
            } else {
                answers[i] = stack.peek();
            }
            // stack.add(currentElem); // add를 써서 그런가?!!!!
            stack.push(currentElem);
        }

        for (int answer: answers) {
            sb.append(answer);
            sb.append(" ");
        }
        System.out.println(sb);
    }
}
